library/Freckle/App/OpenTelemetry.hs

```diff
@@ -1,41 +1,139 @@
 import OpenTelemetry.Trace
   ( HasTracer(..)
+  , Tracer
   , TracerProvider
+  , TracerProviderOptions
   , defaultSpanArguments
   , emptyTracerProviderOptions
+  , makeTracer
+  , tracerOptions
   )
+import qualified OpenTelemetry.Trace.Core as Trace
+  (SpanContext(..), getSpanContext)
+import OpenTelemetry.Trace.Id (Base(Base16), TraceId, traceIdBaseEncodedText)
 import OpenTelemetry.Trace.Monad (MonadTracer(..))
 import qualified OpenTelemetry.Trace.Monad as Trace
-import qualified OpenTelemetry.Trace.Setup as Trace
+import OpenTelemetry.Trace.Setup
 import OpenTelemetry.Trace.Setup.Lens
 
 inSpan :: (MonadUnliftIO m, MonadTracer m, HasCallStack) => Text -> m a -> m a
 inSpan = flip Trace.inSpan defaultSpanArguments
 
+addTraceIdThreadContext :: Middleware
+addTraceIdThreadContext app request respond = do
+  mTraceId <- getCurrentTraceIdHex
+
+  let
+    wrap = case mTraceId of
+      Nothing -> id
+      Just traceId -> withThreadContext ["trace_id" .= traceId]
+
+  wrap $ app request respond
+
 -- | Configure tracing with Id values compatible with AWS X-Ray
 withTracerProviderXRay :: MonadUnliftIO m => (TracerProvider -> m a) -> m a
-withTracerProviderXRay =
-  Trace.withTracerProvider
-    $ (idGeneratorL .~ awsXRayIdGenerator)
-    . (propagatorL <>~ awsXRayContextPropagator)
+withTracerProviderXRay = withTracerProvider setupAWSXRay
+
+-- | 'withTracerProviderXRay' but also configure to log via 'Logger'
+withTracerProviderXRayLogger
+  :: MonadUnliftIO m => Logger -> (TracerProvider -> m a) -> m a
+withTracerProviderXRayLogger logger =
+  withTracerProvider $ setupAWSXRay . setupLogger logger
 
 -- | Configure no tracing
 --
 -- This allows you to still make and set a 'Tracer' and satisfy 'MonadTracer',
 -- but not actually emit any traces.
 --
 withTracerProviderDisabled :: MonadUnliftIO m => (TracerProvider -> m a) -> m a
-withTracerProviderDisabled =
-  Trace.withTracerProvider $ const emptyTracerProviderOptions
+withTracerProviderDisabled = withTracerProvider disableTracing
+
+setupAWSXRay :: TracerProviderOptions -> TracerProviderOptions
+setupAWSXRay =
+  (idGeneratorL .~ awsXRayIdGenerator)
+    . (propagatorL <>~ awsXRayContextPropagator)
+
+setupLogger :: Logger -> (TracerProviderOptions -> TracerProviderOptions)
+setupLogger logger = loggerL .~ go
+ where
+  go l = runLoggerLoggingT logger $ do
+    let (level, message) = fromOTelLog l
+    logOtherNS "hs-opentelemetry-sdk" level message
+
+disableTracing :: TracerProviderOptions -> TracerProviderOptions
+disableTracing = const emptyTracerProviderOptions
+
+-- |
+--
+-- TODO: Use more fields, i.e. metadata
+--
+-- <https://hackage.haskell.org/package/hs-opentelemetry-api-0.0.3.6/docs/OpenTelemetry-Logging-Core.html#t:Log>
+--
+fromOTelLog :: Trace.Log Text -> (LogLevel, Message)
+fromOTelLog = fromOTelSeverity . Trace.severityNumber &&& (:# []) . Trace.body
+
+fromOTelSeverity :: Maybe Int64 -> LogLevel
+fromOTelSeverity = \case
+  Just n | n > 17 -> LevelError
```

Comment:
I always wonder how these severity numbers get picked.

Reply:
God bless open standards,

[The spec](https://opentelemetry.io/docs/reference/specification/logs/data-model/#field-severitynumber) links to [the design discussion](https://docs.google.com/document/d/1ix9_4TQO3o-qyeyNhcOmqAc1MTyr-wnXxxsdWgCMn9c/edit#), which links to a [separate discussion](https://docs.google.com/document/d/1ix9_4TQO3o-qyeyNhcOmqAc1MTyr-wnXxxsdWgCMn9c/edit#) which has a wonderful FAQ for that!

> **Why do we need a range of severity_number values with similar meaning?**
>
> Different logging libraries and logging formats define different sets of possible values for severities. Typically, broadly speaking, the severity values fall in the 6 ranges as defined in this document. However, often there are also more nuanced severity grades in source formats. For example Syslog defines 4 different error severities: (Emergency, Alert, Critical, Error). We need to be able to represent all these severities in the OpenTelemetry data model. The range of severity values makes it possible.
>
> **Why do we have so many values in each range?**
>
> The largest number of values in a range I have found so far is Syslog and Zap. Both have 4 different error levels. There may be formats and libraries that have even more levels so I chose to have 8 levels in each range just to be safe. There does not seem to be downsides to allowing more levels (other than requiring an extra bit to represent).
Why do severity_number values start from 1 and not from 0?
>
> In certain data formats (e.g. ProtoBuf) fields with 0 values are omitted on the wire and there is no way to distinguish omitted and 0 values in receivers. It is efficient to reserve 0 value as an indication of an omitted field (otherwise we would need to record the fact that severity_number field is present or omitted separately).